runtimebp/internal/maxprocs/maxprocs.go

```diff
@@ -0,0 +1,140 @@
+		Help: "Value of the GOMAXPROCS environment variable at startup. 0 if not a number",
+	}, []string{"present"})
+	mEnvCPURequest = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "baseplate_go_env_baseplate_cpu_request",
+		Help: "Value of the BASEPLATE_CPU_REQUEST environment variable at startup. 0 if not a number",
+	}, []string{"present"})
+	mEnvCPURequestScale = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "baseplate_go_env_baseplate_cpu_request_scale",
+		Help: "Value of the BASEPLATE_CPU_REQUEST_SCALE environment variable at startup. 0 if not a number",
+	}, []string{"present"})
+
+	initialGOMAXPROCS = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "baseplate_go_initial_gomaxprocs",
+		Help: "Resolved value of GOMAXPROCS at startup",
+	}, []string{"set_by"})
+
+	_ = promauto.NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "baseplate_go_current_gomaxprocs",
+		Help: "Current value of GOMAXPROCS",
+	}, currentGOMAXPROCS)
+
+	// overridable for tests
+	setWithAutomaxprocs = func() {
+		ubermaxprocs.Set(
+			ubermaxprocs.Min(2),
+			ubermaxprocs.Logger(func(s string, i ...interface{}) {
+				fmt.Fprintf(os.Stderr, s, i...) // contains context
+			}),
+		)
+	}
+)
+
+// Set configures the runtime's GOMAXPROCS using the following heuristic:
+//   1. If $GOMAXPROCS is set, Set relinquishes control to the Go runtime.
+//      This should cause the runtime to respect this value directly.
+//   2. If $BASEPLATE_CPU_REQUEST is unset/invalid, Set relinquishes control to automaxprocs, minimum 2.
+//      See https://pkg.go.dev/go.uber.org/automaxprocs for specific behavior.
+//   3. Otherwise, $BASEPLATE_CPU_REQUEST is multiplied by $BASEPLATE_CPU_REQUEST_SCALE
+//      (or defaultCPURequestScale) to compute the new GOMAXPROCS, minimum 2.
+//
+// Set also exports several metrics to facilitate further tuning/analysis.
+func Set() {
+	envGOMAXPROCS := &floatEnv{key: "GOMAXPROCS", gauge: mEnvGOMAXPROCS}
+	envCPURequest := &floatEnv{key: "BASEPLATE_CPU_REQUEST", gauge: mEnvCPURequest}
+	envCPURequestScale := &floatEnv{key: "BASEPLATE_CPU_REQUEST_SCALE", gauge: mEnvCPURequestScale}
+
+	for _, env := range []*floatEnv{envGOMAXPROCS, envCPURequest, envCPURequestScale} {
+		env.raw, env.present = os.LookupEnv(env.key)
+		env.val, _ = strconv.ParseFloat(env.raw, 64) // error not needed as we check val > 0 and save raw
+		env.gauge.WithLabelValues(strconv.FormatBool(env.present)).Set(env.val)
+	}
+
+	setBy := setByGOMAXPROCS
+	defer func() {
+		initialGOMAXPROCS.WithLabelValues(setBy).Set(currentGOMAXPROCS())
+	}()
+
+	if envGOMAXPROCS.present {
+		return // let Go runtime handle it
+	}
+
+	if !envCPURequest.present {
+		setBy = setByAutomaxprocs
+		setWithAutomaxprocs()
+		return
+	}
+
+	if envCPURequest.val <= 0 {
+		// This should always be valid positive float in infrared-deployed applications.
+		fmt.Fprintf(os.Stderr, "maxprocs: $BASEPLATE_CPU_REQUEST=%q, want positive float. Falling back to Go's default", envCPURequest.raw)
+		return
+	}
```

Comment:
https://github.com/reddit/baseplate.go/pull/547/files#diff-f30af23b7ae4538cec8fe53a3dd75227fb07edf28ca76ccac9031b683b6fb6b9R21-R27

says that this will be handled by `automaxprocs` but it doesn't look like that's what's going on here?

Reply:
You're totally right, good catch – thank you!